From dd8432027a61a615c18882a18142e8e2853239d6 Mon Sep 17 00:00:00 2001
From: Arthur Crepin-Leblond <arthur.crepin@proglove.de>
Date: Sat, 11 Nov 2023 16:57:20 +0100
Subject: [PATCH 1/1] net:bluetooth:hci_event: Refuse connection parameters
 update request

---
 net/bluetooth/hci_event.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/net/bluetooth/hci_event.c b/net/bluetooth/hci_event.c
index e4d8857716eb..ab254388ffcd 100644
--- a/net/bluetooth/hci_event.c
+++ b/net/bluetooth/hci_event.c
@@ -6688,6 +6688,9 @@ static void hci_le_remote_conn_param_req_evt(struct hci_dev *hdev, void *data,
 		return send_conn_param_neg_reply(hdev, handle,
 						 HCI_ERROR_INVALID_LL_PARAMS);
 
+	return send_conn_param_neg_reply(hdev, handle,
+						 HCI_ERROR_INVALID_LL_PARAMS);
+
 	if (hcon->role == HCI_ROLE_MASTER) {
 		struct hci_conn_params *params;
 		u8 store_hint;
-- 
2.42.0

