diff --git a/net/bluetooth/smp.c b/net/bluetooth/smp.c
index cc2f7ca91..6cec48b36 100644
--- a/net/bluetooth/smp.c
+++ b/net/bluetooth/smp.c
@@ -1990,6 +1995,9 @@ static u8 smp_cmd_pairing_rsp(struct l2cap_conn *conn, struct sk_buff *skb)
 	}
 
 	if ((req->auth_req & SMP_AUTH_SC) && (auth & SMP_AUTH_SC))
+    if (key_size != SMP_MAX_ENC_KEY_SIZE)
+        return SMP_ENC_KEY_SIZE;
+
 		set_bit(SMP_FLAG_SC, &smp->flags);
 	else if (conn->hcon->pending_sec_level > BT_SECURITY_HIGH)
 		conn->hcon->pending_sec_level = BT_SECURITY_HIGH;
@@ -2388,12 +2396,17 @@ int smp_conn_security(struct hci_conn *hcon, __u8 sec_level)
 
 	if (hcon->role == HCI_ROLE_MASTER) {
 		struct smp_cmd_pairing cp;
